dev_handler/dev_handler.c

```diff
@@ -1,152 +1,675 @@
 
-			printf("\n");
+    // Cancel this thread only where pthread_testcancel()
+    pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, NULL);
+
+	/* Do work! Continuously read from the device until a complete message can be parsed.
+	 * If it's a PING, update relay->last_received_ping_time
+	 * If it's DEVICE_DATA, write it to shared memory
+	 */
+	// An empty message to parse the received data into
+	message_t* msg = make_empty(MAX_PAYLOAD_SIZE);
+	// An array of empty parameter values to be populated from DeviceData message payloads and written to shared memory
+	param_val_t* vals = malloc(MAX_PARAMS * sizeof(param_val_t));
+	while (1) {
+		// Try to read a message
+		if (receive_message(relay, msg) != 0) {
+            // Message was broken... try to read the next message
+			continue;
+		}
+		if (msg->message_id == PING) {
+            //log_printf(DEBUG, "Got PING from %s", get_device_name(relay->dev_id.type));
+            pthread_mutex_lock(&relay->relay_lock);
+			relay->last_received_ping_time = millis();
+            pthread_mutex_unlock(&relay->relay_lock);
+		} else if (msg->message_id == DEVICE_DATA) {
+		    //log_printf(DEBUG, "Got DEVICE_DATA from %s", get_device_name(relay->dev_id.type));
+			// Get param values from payload
+			parse_device_data(relay->dev_id.type, msg, vals);
+			// Write it to shared memory
+			device_write(relay->shm_dev_idx, DEV_HANDLER, DATA, *((uint32_t *)msg->payload), vals);
+		} else if (msg->message_id == LOG) {
+            log_printf(DEBUG, "[%s]: %s", get_device_name(relay->dev_id.type), msg->payload);
+		} else {
+			log_printf(WARN, "Dropping received message of unexpected type from %s", get_device_name(relay->dev_id.type));
 		}
-		// Wait 5 seconds before
-		count++;
-		sleep(delaySec);
+		// Now that the message is taken care of, clear the message
+		msg->message_id = 0x0;
+		msg->payload_length = 0;
+	    msg->max_payload_length = MAX_PAYLOAD_SIZE;
+		memset(msg->payload, 0, MAX_PAYLOAD_SIZE);
+        pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL);
+		pthread_testcancel(); // Cancellation point
+        pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, NULL);
+	}
+	return NULL;
+}
+
+// ************************************ DEVICE COMMUNICATION ****************************************** //
+
+/**
+ * Helper function for sender()
+ * Serializes, encodes, and sends a message
+ * msg: The message to be sent
+ * return: 0 if successful
+ *         -1 if couldn't write all the bytes
+ */
+int send_message(msg_relay_t* relay, message_t* msg) {
+	int len = calc_max_cobs_msg_length(msg);
+	uint8_t* data = malloc(len);
+	len = message_to_bytes(msg, data, len);
+    int transferred = writen(relay->file_descriptor, data, len);
+    if (transferred != len) {
+        log_printf(WARN, "Sent only %d out of %d bytes to %d\n", transferred, len, get_device_name(relay->dev_id.type));
+    }
+    free(data);
+    //log_printf(DEBUG, "Sent %d bytes: ", len);
+    //print_bytes(data, len);
```

Comment:
should we remove these commented-out print statements? are there any still left in `message.c`?

Reply:
They were crucial print statements used to debug reading and writing raw bytes, but you're right--nowhere in the near future should we need these print statements anymore!